compiler/codegen/cg-CForLoop.cpp

```diff
@@ -72,47 +55,149 @@ static llvm::MDNode* generateLoopMetadata(bool thisLoopParallelAccess,
         anyParallelAccesses = true;
     }
   }
+  return anyParallelAccesses;
+}
 
-  if (anyParallelAccesses) {
-    std::vector<llvm::Metadata*> v;
-    v.push_back(llvm::MDString::get(ctx, "llvm.loop.parallel_accesses"));
+static void addLoopParallelAccess(std::vector<llvm::Metadata*>& args,
+                                  bool thisLoopParallelAccess,
+                                  llvm::MDNode*& accessGroup) {
+  GenInfo* info = gGenInfo;
+  auto &ctx = info->module->getContext();
 
-    // Generate {"llvm.loop.parallel_accesses", group1, group2, ...}
-    // where the groups are any parallel loops we are currently in
-    // (including outer loops to this loop).
-    if (thisLoopParallelAccess) {
-      v.push_back(accessGroup);
-    }
-    for (auto & loopData : info->loopStack) {
-      if (loopData.markMemoryOps) {
-        v.push_back(loopData.accessGroup);
-      }
+  std::vector<llvm::Metadata*> v;
+  v.push_back(llvm::MDString::get(ctx, "llvm.loop.parallel_accesses"));
+
+  // Generate {"llvm.loop.parallel_accesses", group1, group2, ...}
+  // where the groups are any parallel loops we are currently in
+  // (including outer loops to this loop).
+  if (thisLoopParallelAccess) {
+    v.push_back(accessGroup);
+  }
+  for (auto & loopData : info->loopStack) {
+    if (loopData.markMemoryOps) {
+      v.push_back(loopData.accessGroup);
     }
+  }
+
+  llvm::MDNode* parAccesses = llvm::MDNode::get(ctx, v);
+  args.push_back(parAccesses);
+}
+
+static void addLoopRVMetadata(std::vector<llvm::Metadata*>& args) {
+  GenInfo* info = gGenInfo;
+  auto &ctx = info->module->getContext();
+
+  llvm::Constant* one = llvm::ConstantInt::get(llvm::Type::getInt1Ty(ctx), true);
+  llvm::Metadata *loopVectorizeEnable[] = {
+    llvm::MDString::get(ctx, "rv.loop.vectorize.enable"),
+    llvm::ConstantAsMetadata::get(one)
+  };
+
+  args.push_back(llvm::MDNode::get(ctx, loopVectorizeEnable));
+
+  // Note that the Region Vectorizer once required
+  // llvm.loop.vectorize.width but no longer does.
+}
+
+
+
+static llvm::Metadata* constructLLVMMetadata(LLVMMetadataPtr attr) {
```

Comment:
Unless I am mistaken reading the diff, this is something you're adding this PR. It looks to me like there are several cases (where we add `parallel.accesses` or `rv` metadata) where can benefit from a function like this. Is it a reasonably easy refactor to have a `constructLLVMMetadata`-like function that can be used elsewhere during codegen. I think it'd be very valuable to pursue that refactor.

Reply:
I think a good start here is to have a series of functions like `constructLLVMMetdata` that take a string and a value (start with int, bool, char*, and llvm::Metatata*). These can be used to do the `parallel.accesses`, etc. Then a function `constructUserLLVMMetadata` (or a better name, depends on what I come up with for `hasLLVMMetadata`) can call these as helpers.

Or better yet, add a codegen function to `LLVMMetadata`.